Multitooling pod door controllers now picks from a list & auto generates new IDs

SyncIt21 opened the pull request:
## About The Pull Request
Using a multitool on a pod door assembly controller now displays a list of doors to choose from instead of typing an ID from `0->100`. You can only control doors that you have either crafted/linked with an assembly controller It can also auto generate a unique ID for you to assemble new doors

https://github.com/user-attachments/assets/0918f1e3-1f22-4cde-b20e-0eb8bb6ba953

## Why It's Good For The Game
Having to no longer remember the ID of doors you created cause the game now displays that list for you is very convenient. Also generating new unique IDs without player intervention (which results in errors if you can't remember the door IDs you typed before) is faster as well. 
Not to mention currently if you have created 100 doors you run out of unique ID's. Now that's no longer the case

## Changelog
:cl:
qol:  Multitooling a pod door controller now picks from a list of doors available for that player(crafted, set via controller) only and can auto generate new IDs for you
/:cl:

SyncIt21 committed:
Multitooling door controller now picks from a list

00-Steven reviewed (changes requested):
While I think this is very cool, I think it's also fundamentally at odds with shutters/blast doors as protective measures against other players
Or well, this looks like it allows anyone with a multitool and a controller to fuck with *any* closed shutters/blast doors on the station from any point in the game with no further access or knowledge
The roundstart IDs are like that is to avoid letting people open and close them with a random controller, and when you no longer need to know the ID for player-built doors and get to see a handy list-of-IDs-by-area you can't practically use them against other players anymore

So sadly I can't really come up with compromises that keep both this feature *and* their usability for you, so I'm putting this on DNM.

If a hundred unique controller IDs being too low is the issue, I would suggest just increasing that
Beyond that, there should already be things for copying from a controller into a pod door to reduce the amount of memorization or copy-pasting, and if it isn't a thing controllers should be able to copy their IDs to other controllers.
I would be *hesitant* on copying from (open-and-with-open-panel) pod doors into controllers, because it runs back into the mentioned issues, but I could see arguments from it.

SyncIt21 commented:
The solution is simple then. Only list blastdoors who's IDs are numbers I.e not round start blastdoors. 
Would that be okay?

SyncIt21 committed:
only player interacted doors

SyncIt21 commented:
I already mess with round start open pod doors via screw driver and multi tool to change them to my own id (icebox library maint has like 4 blast doors to play with)

You can still do this only difference is you must set the id on the controller and hit the doors with it. I'll implement th multitool on door behavior tomorrow 

> Also will we be able to still set multiple controllers to the same id?

Yes but you must do it before linking the controller to the door.

00-Steven reviewed (commented):
Okay so, I think that addresses the most glaring issues I had with it, so removing the dnm regardless :+1:

On the proposed compromise, I consulted with some more people about potential concerns with this:
1. Removing the manual input makes it impossible to share IDs over PDA/paper/fax/radio/etc with someone, or have someone overhear that, and make a controller based on that. Ideally, that should still be possible.
2. However, a manual input means the current incremental 'unique' ID implementation would allow other players to deduce the IDs in play and copy them. So in addressing 1, that's important to keep in mind.
3. In addition to 2, the current incremental 'unique' ID implementation bases it on IDs used in doors, rather than actually being unique- so two people working on pod doors at the same time might accidentally share an ID despite selecting 'unique'. Ideally, whatever implementation we use resolves that.
4. Looping over all doors to get the IDs used is a bit awkward when we mainly care about IDs used by that player. I'm not 100% sure on the best solution to that, though. The memory system may be worth looking into, but if I recall correctly that has limitations that make this awkward.

Based on that I think it'd be *best* to track the used ID list per player in some way, and still allow a manual ID input. Then tracking the in-use IDs separately globally for generating 'guaranteed unique' IDs.
That said, while I feel it'd be nice to have, working out how to generate 'guaranteed unique IDs that aren't predictable enough to be abusable' is probably not strictly needed if we keep a manual input.

SyncIt21 committed:
generate id on demand

SyncIt21 commented:
Right so addressing some concerns

> Removing the manual input makes it impossible to share IDs over PDA/paper/fax/radio/etc with someone, or have someone overhear that, and make a controller based on that. Ideally, that should still be possible.

You can still examine the controller to get the unique generated ID and share that with players. Its still an incremental integer ID

> However, a manual input means the current incremental 'unique' ID implementation would allow other players to deduce the IDs in play and copy them. So in addressing 1, that's important to keep in mind.

From point 1 there will be no need for manual input of ID via tgui window because examining the controller reveals the ID

> In addition to 2, the current incremental 'unique' ID implementation bases it on IDs used in doors, rather than actually being unique- so two people working on pod doors at the same time might accidentally share an ID despite selecting 'unique'. Ideally, whatever implementation we use resolves that.

So this problem occurs because the unique ID is generated from a list of IDs that have been registered on the existing doors. If an ID has been generated but hasn't been registered to on a door yet then that ID will not be unique. Now i have fixed that. The random ID is now generated at the time of linking the controller with the door(by hitting it) & not at the time of using the multitool so we now have full guarantee that all IDs will be unique.

Now addressing an subpoint to this problem
> I already mess with round start open pod doors via screw driver and multi tool to change them to my own id (icebox library maint has like 4 blast doors to play with)
> 
> and i would like this to be preserved

@moocowswag this behaviour cannot be preserved because of the risk of players entering duplicate ID via the tgui window. To get fully unique IDs only the code can ensure that and not player input so sadly this behaviour has to be discarded now on 2nd thought

> Looping over all doors to get the IDs used is a bit awkward when we mainly care about IDs used by that player. I'm not 100% sure on the best solution to that, though. The memory system may be worth looking into, but if I recall correctly that has limitations that make this awkward.

Yeah i have no solution?(Maybe not a big deal?) to this. Looping over all doors in game is the only way to ensure no door gets the same ID. Creating a global/static list just to keep track of player inputed IDs is just additional overhead and gives us the same outcome anyway

New video in the description

SyncIt21 committed:
no default id

SyncIt21 committed:
actual fixes

SyncIt21 committed:
context

SyncIt21 committed:
ok

SyncIt21 committed:
Merge branch 'master' into IDControlPick

SyncIt21 commented:
What the heck? @vinylspiders your label PR something broke with your label PR.

DNM was removed from my PR and yet syncing the branch with master added it back.

@00-Steven can you please remove the label and merge the PR if you can?

SyncIt21 committed:
Merge remote-tracking branch 'upstream/master' into IDControlPick

SyncIt21 commented:
@vinylspiders your PR #92614 didn't work. The do not merge label is still there and was not removed

SyncIt21 committed:
Merge branch 'master' into IDControlPick

SyncIt21 committed:
Merge remote-tracking branch 'upstream/master' into IDControlPick

SyncIt21 committed:
Merge branch 'master' into IDControlPick

SyncIt21 committed:
Merge remote-tracking branch 'upstream/master' into IDControlPick

SyncIt21 committed:
Merge branch 'master' into IDControlPick

SyncIt21 committed:
Just put this in here

SyncIt21 committed:
put this back

SyncIt21 committed:
Merge remote-tracking branch 'upstream/master' into IDControlPick

SyncIt21 committed:
Merge remote-tracking branch 'upstream/master' into IDControlPick

SyncIt21 commented:
@Ghommie can you please merge this PR? I've pinged other maintainers but no one has responded.
It's more than a month old now

Acer202 commented:
Did this break ignition controllers? I can't seem to connect one with an ID after this.
